test: update test and fixed conflict test deploy

madeinfree opened the pull request:
Maybe ready.

- fixed conflict deploy, use beforeEach to instead of before

# TODO LIST

- [x] Check ( Variable or Constant )
  - [x] maxMintSupply
  - [x] mintPrice
  - [x] refundPeriod
  - [x] maxUserMintAmount
  - [x] refundEndTime test
  - [x] refundGuaranteeActive test
- [x] Merkle Proof test
  - [x] add merkle tree presale test
- [x] set test
  - [x] add setRefundAddress test
  - [x] add setMerkleRoot test
  - [x] add setBaseURI test
- [x] toggle test
  - [x] add toggleRefundCountdown test
  - [x] add togglePresaleStatus test
  - [x] add togglePublicSaleStatus test
- [x] ownerMint test
  - [x] `Max mint supply reached`
  - [x] owner mint
- [x] publicSaleMint test
  - [x] `Over mint limit`
  - [x] `Max mint supply reached`
  - [x] `Public sale is not active`
  - [x] `Not enough eth sent`
- [x] preSaleMint test
  - [x] `Presale is not active`
  - [x] `Value`
  - [x] `Max amount`
  - [x] `Max mint supply`
- [x] refund test
  - [x] `Refund expired`
  - [x] `Not token owner`
  - [x] `Already refunded`
  - [x] `Freely minted NFTs cannot be refunded`
  - [x] hasRefunded test
- [x] withdraw test
  - [x] `Refund period not over`
  - [x] withdraw after refundEndTime

madeinfree commented:
Hi, @elie222 I think all the test coverage is done, if any suggestions please let me know, I'd happy to help.
and this PR install the hardhat gas-report plugin and add a script for it,  thank you!

```
> npx hardhat test

  ERC721RExample
    ✔ [Check] Check maxMintSupply = 8000
    ✔ [Check] Check mintPrice = 0.1
    ✔ [Check] Check refundPeriod 3888000
    ✔ [Check] Check maxUserMintAmount 5
    ✔ [Check] Check refundEndTime is same with block timestamp in first deploy
    ✔ [Check] Check refundGuaranteeActive = true
    ✔ Should be able to deploy
    ✔ Should be able to mint and request a refund (76ms)
    ✔ [OwnerMint] Should able to mint
    ✔ [PublicMint:Revert] Should not be able to mint when `Public sale is not active` (54ms)
    ✔ [PublicMint:Revert] Should not be able to mint when `Not enough eth sent`
    ✔ [PublicMint:Revert] Should not be able to mint when `Max mint supply reached`
    ✔ [PublicMint:Revert] Should not be able to mint when `Over mint limit`
    ✔ [PreSaleMint:Revert] Should not presale mint when `Not on allow list`
    ✔ [PreSaleMint] Should presale mint merkle tree with valid leaf (43ms)
    ✔ [PreSaleMint:Revert] Should not be mint when `Presale is not active`
    ✔ [PreSaleMint:Revert] Should not be mint when `Value` not enough
    ✔ [PreSaleMint:Revert] Should not be mint when `Max amount` (48ms)
    ✔ [PreSaleMint:Revert] Should not be mint when `Max mint supply`
    ✔ [Refund] Check hasRefunded store correct tokenId (51ms)
    ✔ [Refund:Revert] Should not be refunded when `Not token owner`
    ✔ [Refund:Revert] `Freely minted NFTs cannot be refunded`
    ✔ [Refund:Revert] NFT cannot be refunded twice `Already refunded` (73ms)
    ✔ [Refund] NFT refund should in 45 days (44ms)
    ✔ [Refund:Revert] NFT refund expired after 45 days `Refund expired`
    ✔ [Owner:Revert] Owner should not be able to mint when `Max mint supply reached`
    ✔ [Owner:Revert] Owner can not withdraw when `Refund period not over`
    ✔ [Owner] Owner can withdraw after refundEndTime
    ✔ [Toggle] Owner can call toggleRefundCountdown and refundEndTime add `refundPeriod` days.
    ✔ [Toggle] Owner can call togglePresaleStatus
    ✔ [Toggle] Owner can call togglePublicSaleStatus
    ✔ [Setter] Owner can call setRefundAddress
    ✔ [Setter] Owner can call setMerkleRoot

  33 passing (4s)
```

elie222 commented on test/test.js:
```diff
@@ -63,9 +277,149 @@ describe("ERC721RExample", function () {
+
+  it("[Setter] Owner can call setMerkleRoot", async function () {
+    await erc721RExample.setMerkleRoot(
+      "0x070e8db97b197cc0e4a1790c5e6c3667bab32d733db7f815fbe84f5824c7168d"
```
for merkle roots we can do something along these lines (taken from some other code we have that isn't public yet):

when using a count:

```
  // based on: https://github.com/steve-ng/merkle-airdrop/blob/main/test/MerkleDistributor.ts#L9
    users = [
      { address: owner.address, amount: 10 },
      { address: account2.address, amount: 15 },
      { address: account3.address, amount: 20 },
      { address: account4.address, amount: 30 },
    ];
    elements = users.map((x) =>
      utils.solidityKeccak256(["address", "uint256"], [x.address, x.amount])
    );
    
    // Merkle tree util functions
function getTree(elements) {
  const tree = new MerkleTree(elements, keccak256, { sort: true });
  const root = tree.getHexRoot();
  return { tree, root };
}

function getProof(tree, leaf) {
  return tree.getHexProof(leaf);
}

    const leaf = elements[3];
    const amount = users[3].amount;
    const { root, tree } = getTree(elements);
    const proof = getProof(tree, leaf);

    await cryptoFightersPotion.setMerkleRoot(root);
```

when using just addresses you can skip this:

```
utils.solidityKeccak256(["address", "uint256"], [x.address, x.amount])
```

and just hash the address. something like this:

```
utils.solidityKeccak256(x.address)
```

elie222 reviewed (changes requested):
a note about harded coded addresses and merkle tree roots/proofs

elie222 commented:
Awesome PR! Really helpful :)

madeinfree commented:
@elie222 thank you for quick reply and suggestions, I'll update it soon.

madeinfree commented:
@elie222 I updated the hardcode test and using the merkletreejs, and I move merkletree function to `scripts` folder, what do you think ? thank you for your reply ~

elie222 commented on test/test.js:
```diff
@@ -1,20 +1,34 @@
 const MINT_PRICE = "0.1";
+const MAX_MINT_SUPPLY = 8000;
+const MAX_USER_MINT_AMOUNT = 5;
+const FORTY_FIVE_DAYS = 24 * 60 * 60 * 45;
```
would probably name this `REFUND_PERIOD` or similar

elie222 commented on test/test.js:
```diff
@@ -23,6 +37,38 @@ describe("ERC721RExample", function () {
+    expect(await erc721RExample.refundGuaranteeActive()).to.be.true;
+  });
+
   it("Should be able to deploy", async function () {});
```
i added this but let's remove it

elie222 commented on test/test.js:
```diff
@@ -40,15 +86,195 @@ describe("ERC721RExample", function () {
+    expect(await erc721RExample.ownerOf(0)).to.be.equal(owner.address);
+  });
+
+  /**
+   * PublicMint Test
+   */
```
wondering if we need all these comments. seems like a `describe` block is what you want perhaps?

elie222 replied:
also at the start of every test name. i'd potentially just start each test with `should x` and use describe blocks to group instead

elie222 commented on test/test.js:
```diff
@@ -40,15 +86,195 @@ describe("ERC721RExample", function () {
+    await expect(
+      erc721RExample.preSaleMint(
+        1,
+        ["0x00314e565e0574cb412563df634608d76f5c59d9f817e85966100ec1d48005c0"],
```
we still have some hardcoded addresses

elie222 commented:
Left a few more comments. Some minor stylistic issues

madeinfree commented:
@elie222 thank you, I updated and checked all test hardcode address, merkle tree and some stylistic issue,
rebase the main branch and fixed some unnecessary code.

elie222 commented:
thanks. you didn't want to use describe blocks instead of starting each test with something like this:
`[PublicMint:Revert]`?

elie222 commented:
I just merged in in any case but maybe the describe blocks approach would be cleaner naming wise

madeinfree commented:
Ok, I'll update later and create a new PR!
